[SE-0365] Fix issue where implicit self was unexpectedly disallowed in nested closures

calda opened the pull request:
This PR fixes https://github.com/apple/swift/issues/70089 and https://github.com/apple/swift/issues/69911. Previously, the following sample code unexpectedly failed to compile:

```swift
class Test {
  var x: Int = 0

  func f() {
    doVoidStuff { [weak self] in
      guard let self else { return }
            
      doVoidStuff { [self] in
        x += 1 // error: reference to property 'x' in closure requires explicit use of 'self' to make capture semantics explicit
      }

      doVoidStuffNonEscaping {
        x += 1 // error: reference to property 'x' in closure requires explicit use of 'self' to make capture semantics explicit
      }
    }
  }
}

func doVoidStuff(_: @escaping () -> Void) {}
func doVoidStuffNonEscaping(_: () -> Void) {}
```

Both of these examples are expected to compile based on the behavior described [here](https://github.com/apple/swift-evolution/blob/main/proposals/0365-implicit-self-weak-capture.md#nested-closures) in SE-0365:

> ```swift
> // Also allowed:
> couldCauseRetainCycle { [weak self] in
>   guard let self else { return }
>   foo()
> 
>   couldCauseRetainCycle { [self] in
>     bar()
>   }
> }
> ```

While SE-0365 doesn't explicitly address the behavior of non-escaping closures, it does say that the behavior should follow the same rules as [SE-0269]([SE-0269](https://github.com/apple/swift-evolution/blob/main/proposals/0269-implicit-self-explicit-capture.md)), which does allow implicit self in nested non-escaping closures:

```swift
doVoidStuff { [self] in
  doVoidStuffNonEscaping {
    x += 1 // ok
  }
}
```

Now we allow implicit self in these nested closures, as long as either: 
 - the closure is non-escaping, or
 - the closure explicitly captures self in a valid way

Please review: @xedin @jumhyn (hi again, thanks!)

Fixes https://github.com/apple/swift/issues/70089. Fixes https://github.com/apple/swift/issues/69911.

calda committed:
[SE-0365] Fix issue where implicit self was unexpectedly disallowed in nested closures

calda commented:
Hi @xedin could you review this bug fix? Thanks!

calda commented:
Still looking for a review for this bug fix, @xedin could you help?

Also tagging other folks who GitHub lists as reviewers for these files: @hborla @slavapestov 

Thanks!

xedin commented:
Sorry I missed a notification about this, I'll take a look next week!

calda commented:
Thank you! 🙇🏻

calda commented:
I noticed the current implementation in this PR has an issue where it unexpectedly allows these examples to compile even though they shouldn't:

```swift
doVoidStuff { [self = MyTestClass()] in
  doVoidStuff { [self] in
    x += 1
  }
}
```

```swift
doVoidStuff { [weak self] in
  guard let self = self ?? MyTestClass.staticOptional else { return }
  doVoidStuff { [self] in
    x += 1
  }
}
```

I have a rough change put together which fixes this issue. I plan on cleaning it up tomorrow and updating this PR. I'm also taking a look at https://github.com/apple/swift/issues/67466 since it's relevant (which is how I ended up noticing this edge case).

calda commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -492,11 +482,29 @@ ValueDecl *UnqualifiedLookupFactory::ResultFinderForTypeContext::lookupBaseDecl(
+  // are never allowed to rebind self.
+  if (auto var = dyn_cast<VarDecl>(selfDecl)) {
+    if (!(var->isCaptureList() || var->getParentPatternStmt())) {
+      return nullptr;
```
This is the core of the change. To allow an unwrapped `guard let self else { return }` binding to be used as implicit self, the closure has to use the `ASTScope::lookupSingleLocalDecl` lookup behavior for implicit self. Previously we only did this for closures with a `[weak self]` capture, even though any closure could need this behavior.

The majority of the rest of the changes in the PR are a rework required to reject invalid code like https://github.com/apple/swift/pull/70575#issuecomment-1951685335, which was previously rejected thanks to this problematic behavior here where the closures didn't use the newer lookup behavior.

xedin commented:
Sorry about the delay, @hamishknight is going to take a look at this one.

hamishknight commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1680,34 +1656,174 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+          return false;
+        }
+
+        // If the closure has an explicit capture using the name `self` that
+        // actually refers to some other variable (e.g. `[self = "hello"]`)
+        // then implicit self is not allowed.
+        if (capturedSelfDecl && !capturedSelfDecl->isSelfParamCapture()) {
```
Additionally, it seems like this is currently rejected:

```swift
class C {
  func foo() {
    let _ = { [weak self] in
      let _ = { [self] in
        guard let self else { return }
        foo()
      }
    }
  }
}
```

calda replied:
> Do we need to be checking this for the `weak` case too? Seems like we currently allow e.g:
> 
> ```swift
> class C {
>   func bar() {}
>   func foo(_ x: C) {
>     let fn = { [weak self = x] in
>       guard let self else { return }
>       bar()
>     }
>   }
> }
> ```

Ahhh thanks for noticing this! You're right that this is unexpectedly permitted in Swift 5.10. I'm surprised we missed this in the original implementation of SE-0365. 

I added a check for this, so now we reject it as expected. Since this is source breaking, it only emits a warning in Swift 5 mode, and emits an error in Swift 6 mode.

> Additionally, it seems like this is currently rejected:
> 
> ```swift
> class C {
>   func foo() {
>     let _ = { [weak self] in
>       let _ = { [self] in
>         guard let self else { return }
>         foo()
>       }
>     }
>   }
> }
> ```

This sort of case isn't explicitly covered by the text of SE-0365. All of the examples of supported patterns put the `guard let self else { return }` condition in the outer closure rather than the inner closure. 

I could see it making sense to support this, but since it isn't discussed in the proposal I think it's best to leave it as-is for now. In the current change on this branch we still reject it, like today in Swift 5.10.

hamishknight replied:
Fair enough, I'm happy to leave that as future work

hamishknight reviewed (commented):
Thanks for fixing this!

calda committed:
Fix edge cases related to nested autoclosures, invalid weak self unwrapping

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1727,18 +1843,30 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       return conditionalStmt->rebindsSelf(Ctx, /*requireLoadExpr*/ true);
```
> It's not quite related to this PR, but I'm not sure I understand the `requireLoadExpr` parameter of rebindsSelf. 

Thanks for noticing this, you're right that this check was previously over simplistic and unexpectedly allowed this example to compile:

```swift
var `self` = self ?? .somethingElse
guard let self = self else { return }
method() // this will be allowed
```

I updated the check to properly reject this sort of case. This change is source-breaking, so it emits a warning in Swift 5 mode and an error in Swift 6 mode.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1680,34 +1656,174 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+          outerSelfDecl = ASTScope::lookupSingleLocalDecl(
+              inClosure->getParentSourceFile(), DeclName(ctx.Id_self),
+              capturedSelfDecl->getLoc());
+        }
```
This seems like a good suggestion so me, but I'm not sure how to do this:

> could we check the initializer expression for the capture?

I checked all of the properties on the `VarDecl`, and none of them returned anything useful. For example it sounds like you could be referring to something like `VarDecl::getParentInitializer`, but that returns `nullptr` here.

If there's something specific you had in mind that I missed, could you share a code pointer?

hamishknight replied:
You should be able to retrieve it by iterating over the capture list entries and finding the one that corresponds to the var, from there you should be able to query the PatternBindingDecl for the initializer (using similar logic to `isParentInitialized`, you can query the entry index, then get the expr).

calda replied:
Thanks for the tip, that was really helpful!

I removed the `ASTScope::lookupSingleLocalDecl` call in favor of this approach (accessing the `PatternBindingDecl` via the capture list, and then retrieving the relevant `DeclRefExpr` via `PatternBindingDecl::getInit`): https://github.com/apple/swift/pull/70575/commits/97178b24946c34d97e90b771b85f88f23d9560cd

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1727,18 +1843,30 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      // redefine what 'self' means, even when nested inside a closure.
+      if (dc->getContextKind() == DeclContextKind::GenericTypeDecl) {
+        return nullptr;
+      }
```
Thanks for the suggestion, good idea. I consolidated this into a shared `DeclContext:: getInnermostClosureForSelfCapture()` method, and updated the check to use `dc->isTypeContext()`.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1680,34 +1656,174 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+          // Implicit self is always allowed in non-escaping autoclosures
+          return true;
+        }
+      }
```
Good thinking! I hadn't thought to test nested autoclosures.

I added a ton of tests for nested autoclosures and fixed a few issues. Now this example is rejected as expected.

calda commented:
Thank you for the very thorough review @hamishknight! I especially appreciate that you noticed and called out several existing bugs in the previous implementation. Great that we caught those in time for Swift 6, considering the fix is source-breaking.

I updated the PR to fix all of the issues you mentioned, and added even more tests.

calda committed:
Improve quality of 'capture self explicitly to enable implicit self in this closure' and 'variable other than self captured here under the name self does not enable implicit self' diagnostics

calda commented:
I also just pushed some final cleanup that improves the quality of some of the informational diagnostics.

Before we would sometimes emit confusing `'variable other than self captured here under the name self does not enable implicit self'` diagnostics even for closures with a `[self]` capture, and would sometimes emit `capture 'self' explicitly to enable implicit 'self' in this closure` diagnostics in cases where adding an explicit self capture wouldn't actually fix the issue (because some parent closure was invalid). Both of those edge cases are fixed and improved now 👍🏻

hamishknight commented:
Thanks! I'll try and take a look tomorrow

calda committed:
Look up the outer self decl via the capture's parent initializer instead of ASTScope::lookupSingleLocalDecl

hamishknight commented:
Sorry, I haven't forgotten about this, I'll try and take a look today or tomorrow

calda committed:
Merge branch 'main' into cal--fix-70089

hamishknight commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1874,23 +2114,111 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       };
 
       SourceLoc memberLoc = SourceLoc();
+      const ValueDecl *selfDecl;
```
```suggestion
      const ValueDecl *selfDecl = nullptr;
```

hamishknight commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -430,11 +399,29 @@ ValueDecl *UnqualifiedLookupFactory::lookupBaseDecl(const DeclContext *baseDC) c
+  //
+  // Other types of rebindings, like an arbitrary "let `self` = foo",
+  // are never allowed to rebind self.
+  if (auto var = dyn_cast<VarDecl>(selfDecl)) {
+    if (!(var->isCaptureList() || var->getParentPatternStmt())) {
+      return nullptr;
```
One interesting question is also how we ought to handle things like this:

```swift
struct S {
  func bar() {
    let fn = { [self = S()] in
      bar()
    }
  }
}
```

Today this ignores the bound `self` and uses the `self` of the method. With your change, we will switch to the capture list, but not diagnose because it's not a class type. I wonder if the validation of the `self` rebinding ought to be its own diagnostic that applies generally to any implicit use of `self` that isn't a rebinding of the `self` of the context, which we could stage in via a warning.

calda replied:
> One interesting question is also how we ought to handle things like this:
> 
> ```swift
> struct S {
>   func bar() {
>     let fn = { [self = S()] in
>       bar()
>     }
>   }
> }
> ```
> 

Thanks for thinking of this, I hadn't considered that case before. I added an error for this in Swift 6 mode, staged in with a warning in Swift 5 mode.

> Just realized we accept arbitrary non-VarDecl results here, e.g this becomes valid:
> 
> ```swift
> class C {
>   func bar() {
>     struct `self` {
>       static func bar() {}
>     }
>     let fn = {
>       bar()
>     }
>   }
> }
> ```
> 
> Should we be rejecting non-VarDecl results here?

Good thinking. I tested this and is was rejected as expected, but I went ahead and also rejected non-VarDecls here to be defensive, and added tests for this sort of case.

hamishknight reviewed (commented):
So sorry for the belated review! I have a few more comments, but otherwise this is looking pretty good I think

calda committed:
Always validate self capture

calda commented on lib/AST/Stmt.cpp:
```diff
@@ -578,8 +587,22 @@ bool StmtConditionElement::rebindsSelf(ASTContext &Ctx,
-    return decl && decl->hasName() ? decl->getName().isSimpleName(Ctx.Id_self)
-                                   : false;
+
+    bool definedInCaptureList = false;
+    if (auto varDecl = dyn_cast_or_null<VarDecl>(DRE->getDecl())) {
+      definedInCaptureList = varDecl->isCaptureList();
+    }
+
+    if (requiresCaptureListRef && !definedInCaptureList) {
+      return false;
+    }
```
Good thinking, this unexpectedly compiled before. Checking that the outermost self capture is a capture of the self parameter fixes this nicely.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1757,65 +1733,324 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+              !hasValidSelfRebinding(parentConditionalStmt(selfDecl),
+                                     inClosure)) {
+            return false;
+          }
```
Thanks for noticing, the tests still pass without this so I removed it.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1757,65 +1733,324 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+
+        auto outerClosureContainsSelfDecl =
+            outerClosure->getSourceRange().contains(
+                outerSelfDecl->getLoc(true));
```
Ahh yes this is much better, thank you! That worked perfectly.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1874,23 +2114,111 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       auto shouldOnlyWarn = [&](Expr *selfRef) {
+        // We know that isImplicitSelfParamUseLikelyToCauseCycle is true,
+        // which means all these casts are valid.
+        auto selfDecl = cast<VarDecl>(cast<DeclRefExpr>(selfRef)->getDecl());
```
Removed these assumptions added standard checks that these are nonnull etc

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1757,65 +1733,324 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      const ValueDecl *outerSelfDecl = selfDecl;
+      if (auto closureExpr = dyn_cast<ClosureExpr>(inClosure)) {
+        if (auto capturedSelfDecl = closureExpr->getCapturedSelfDecl()) {
+          // Retrieve the outer decl that the self capture refers to.
+          outerSelfDecl = getParentInitializerDecl(capturedSelfDecl);
```
This check happens below when we call `selfDeclAllowsImplicitSelf(outerSelfDecl, captureType, outerClosure)` with this decl and its containing closure

calda commented:
Thanks again for the review @hamishknight! I pushed a change that fixes the edge cases you pointed out, plus your other suggestions.

hamishknight reviewed (commented):
Sorry for dropping the ball on this again, I have a few more comments (mostly minor, though I came up with a couple of cases around `shouldOnlyWarn` I think we probably ought to fix), but otherwise this looks good to me

hamishknight commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1874,51 +2132,177 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+          }
+        }
+
+        // If the self decl refers to a weak self unwrapping
+        // condition in some parent closure, then there is no
+        // source-compatibility requirement to avoid an error.
+        if (hasValidSelfRebinding(parentConditionalStmt(selfDecl), ACE)) {
+          return false;
         }
```
FWIW it looks like the only case that fails with this check removed is:

```swift
    doVoidStuff { [weak self] in      
      guard let self else { return }

      doVoidStuffNonEscaping {
        withEscapingAutoclosure(bar()) // expected-error {{call to method 'bar' in closure requires explicit use of 'self' to make capture semantics explicit}} expected-note {{reference 'self.' explicitly}}
      }
    }
```

But that isn't really to do with the binding, that's because we would have previously referenced the `self` of the function itself. Maybe the carve out should be specific to escaping autoclosures?

calda replied:
Good catch! Both of these should be warnings rather than errors. Updated!

calda committed:
Improve 'shouldOnlyWarn' logic

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1874,51 +2132,177 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+              return false;
+            }
+          }
+        }
```
You're right, good thinking. I was able to remove this check. It must have been superseded by one of the other checks below, possibly the `parentClosureDisallowingImplicitSelf` check when I remember came later.

calda commented on test/expr/closure/closures.swift:
```diff
@@ -1168,3 +1211,470 @@ final class TestGithubIssue64757 {
+      }
+      
+      doVoidStuff {
+        withEscapingAutoclosure(bar()) // expected-error {{call to method 'bar' in closure requires explicit use of 'self' to make capture semantics explicit}} expected-note {{reference 'self.' explicitly}}
```
Good catch, fixed this to stay a warning instead of an error.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1706,116 +1706,369 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      } while (true);
+    }
+
+    static bool
+    hasValidSelfRebinding(const LabeledConditionalStmt *conditionalStmt,
+                          const AbstractClosureExpr *inClosure) {
```
This is a static method so I don't see another `ASTContext` we can access. In the other related static helpers we also access the `ASTContext` through the `AbstractClosureExpr`. lmk if I missed something though and I can update this.

hamishknight replied:
Ah sorry, missed that this was `static`

calda committed:
Update DeclContext::getInnermostClosureForSelfCapture to return a ClosureExpr

calda commented on lib/AST/DeclContext.cpp:
```diff
@@ -263,6 +263,25 @@ DeclContext *DeclContext::getInnermostSkippedFunctionContext() {
   return nullptr;
 }
 
+DeclContext *DeclContext::getInnermostClosureForSelfCapture() {
```
Updated -- when I first added this I tried to import `Expr.h`, which didn't work because it was a circular reference. Just noticed the `namespace swift {` that we can use for a forward declaration.

calda commented:
Thanks again @hamishknight! Glad we could iron out more of these edge cases -- updated the PR to incorporate your suggestions.

calda committed:
Simplify autoclosure check in shouldOnlyWarn

hamishknight commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1706,116 +1713,362 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      }
+
+      if (auto autoclosure = dyn_cast<AutoClosureExpr>(inClosure)) {
+        // Implicit self is always allowed in autoclosure thunks generated
+        // during type checking. An example of this is when storing an instance
+        // method as a closure (e.g. `let closure = someInstanceMethodOnSelf`).
+        auto isThunk =
+            autoclosure->getThunkKind() != AutoClosureExpr::Kind::None;
+        if (isThunk) {
+          return true;
+        }
```
Sorry I didn't catch this before (and sorry to keep piling on more counterexamples 😅), but I just noticed that this would allow the following to compile, which is currently not allowed in 5.10 (and seems like it should remain illegal):

```swift
class C {
  func bar() -> Int { 0 }
  func foo() async {
    let _ = {
      async let x = bar()
    }
  }
}
```

I think we still need to check parent closures in this case.

calda replied:
Good catch, thanks! Fixed.

slavapestov commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1874,51 +2131,163 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       // Until Swift 6, only emit a warning when we get this with an
       // explicit capture, since we used to not diagnose this at all.
       auto shouldOnlyWarn = [&](Expr *selfRef) {
+        auto DRE = dyn_cast_or_null<DeclRefExpr>(selfRef);
```
This local function is over a hundred lines long, and it's inside a method that's even longer. I feel like this 'implicit self' diagnostic does not carry its weight, because it introduces a lot of new bookkeeping in name lookup, and the diagnostic logic here is also quite complex. Is there any way we can simplify the rule? Is the diagnostic worth having at all?

slavapestov replied:
For example, here's an idea: instead of "shouldOnlyWarn", why not make it always a warning? Then you can delete some code...

calda replied:
In terms of whether or not the diagnostic should exist at all: it implements functionality defined in [SE-0269](https://github.com/apple/swift-evolution/blob/main/proposals/0269-implicit-self-explicit-capture.md) and [SE-0365](https://github.com/apple/swift-evolution/blob/main/proposals/0365-implicit-self-weak-capture.md), which meaningfully improve the ergonomics of working with closures, and have been accepted for some time (but with bugs in the existing implementations).

slavapestov replied:
> If we wanted to simplify the code we could deprioritize some of the edge cases and accept these downgraded warnings in more cases in Swift 5 mode.

My overall concern is that the existence of these edge cases suggests something isn't being encoded or analyzed properly and we're probably missing a simpler model that would produce the same diagnostics. A reasonable way to start is by splitting up some of these longer functions so that it's easier to tell what's going on. I also find that nested local functions can become unreadable when they're longer than, say, 20 lines of code each. It might even be worth moving some stuff out of MiscDiagnostics.cpp into a new file.

I'm fully aware of the fact that MiscDiagnostics.cpp is already a huge tangled mess -- after all, we're diagnosing a bunch of stuff that just doesn't fit anywhere else -- but that's all the more reason to tidy things up. I want to say I definitely appreciate your efforts, and it would be great to chip away at the complexity a little bit more.

You can do any of this in a follow-up PR since you've been iterating on this one for a while already.

Also FWIW, I recently added this FIXME in `diagnoseImplicitSelfUseInClosure()`:

```
      // FIXME: This is happening too early, because closure->getType() isn't set.
      if (auto *closure = dyn_cast<AbstractClosureExpr>(DC))
        if (closure->getType())
          if (DiagnoseWalker::isClosureRequiringSelfQualification(closure, ctx))
            ACE = const_cast<AbstractClosureExpr *>(closure);
```

The getType() check was hidden behind a utility function before, and it would silently skip the isClosureRequiringSelfQualification() check if the closure hadn't been type checked yet. This suggests the visitor is looking at too many things, or looking at things in the wrong order. If you change the check into an assertion you'll see the problem.

slavapestov replied:
> without unintentionally downgrading any existing errors to warnings.

This is a noble goal but I wouldn't prioritize it over all else. It's always safer to just downgrade to a warning.

calda replied:
Thanks -- for this PR I'll at least reorganize the code to be easier to follow, and do a pass over each of the checks in `shouldOnlyWarn` to see if any are good candidates to be removed for simplicity.

calda replied:
Thanks for the advice @slavapestov! I refactored and simplified the `shouldOnlyWarn` logic in https://github.com/apple/swift/pull/70575/commits/d17bc35a93a6b9073162ceea80704324e8beeec4. The code is much better organized now, so should be much easier to follow.

I reviewed each of the individual checks, and was able to remove a few. The checks that remain are all targeted and well motivated (e.g. required to preserve source compatibility).

calda committed:
Fix unexpected error by simplifying outermost capture validation

calda committed:
Refactor and simplify shouldOnlyWarn logic

calda commented:
Thanks for catching so many edge cases @hamishknight! Updated to fix those unexpected new errors, and also cleaned up the implementation of `shouldOnlyWarn`.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1874,51 +2131,163 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+        // This doesn't apply to autoclosures.
+        if (!isa<AutoClosureExpr>(ACE)) {
+          if (auto invalidParentClosure =
+                  parentClosureDisallowingImplicitSelf(selfDecl, ty, ACE)) {
```
Thanks for noticing this! Added a test case for this. It's no longer an issue after simplifying the `shouldOnlyWarn` logic.

calda commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1874,51 +2131,163 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+        //     }
+        //   }
+        //
+        return !selfDecl->isSelfParameter() && !closureHasExplicitSelfCapture;
```
You're right, thanks -- updated the `closureHasExplicitSelfCapture` check to only apply to escaping closures.

calda committed:
Fix validation for async let thunks

calda committed:
Merge tag 'swift-DEVELOPMENT-SNAPSHOT-2024-05-01-a' of github.com:apple/swift into cal--fix-70089

Tag build swift-DEVELOPMENT-SNAPSHOT-2024-05-01-a

calda committed:
Remove unused validateOutermostCapture param

calda committed:
Fix outdated usage of isKnownNoEscape

slavapestov commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1708,124 +1715,365 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
+      if (auto autoclosure = dyn_cast<AutoClosureExpr>(inClosure)) {
+        // Implicit self is always allowed in autoclosure thunks generated
+        // during type checking. An example of this is when storing an instance
+        // method as a closure (e.g. `let closure = someInstanceMethodOnSelf`).
```
I feel like this goes against the spirit of the proposal. Wouldn't you want to write `let closure = self.someInstanceMethod` for consistency and to make the self capture explicit? Or is this another historical quirk?

calda replied:
Here's an example that fails to compile without this:

```swift
class Test {
  func foo() {
    let closure = someInstanceMethod
    closure()
  }
  
  func someInstanceMethod() { }
}
```

Allowing this to compile seems reasonable to me. It's also required for source compatibility.

However, if you instead access `someInstanceMethod` inside a closure, explicit self is required as you'd definitely expect:

```swift
class Test {
  func foo() {
    _ = {
      let closure = someInstanceMethod // error: error: unexpected error produced: implicit use of 'self' in closure
      closure()
    }
  }
  
  func someInstanceMethod() { }
}
```

I double-checked locally, the reason this works is that the implicit self decl is validated from both the context of the outer ClosureExpr and the AutoClosureExpr. This happens because the AST is `(ClosureExpr (DotSyntaxCallExpr (AutoClosureExpr (DeclRefExpr))))`, and both the DSCE and DRE are validated separately within the context of their containing closure using different rules.

Jumhyn replied:
Yeah, doing the `self` capture analysis for bound, unapplied method references was discussed back in the SE-0269 days but never made its way into being an actual proposal. Not sure there'd be appetite for introducing _additional_ cases where implicit `self` would be disallowed at this point, though I'd probably count myself in favor :)

slavapestov reviewed (approved):
Thanks for the cleanup!

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

calda commented:
Tests passed!

The three failures in the source compatibility suite look unrelated to me: https://ci.swift.org/job/swift-PR-source-compat-suite-macos/1585/testReport

hamishknight reviewed (approved):
Thanks for sticking with this! LGTM

calda commented:
Thanks everyone! 🙇🏻 I also posted a PR for the Swift 6.0 release branch: https://github.com/apple/swift/pull/73482
